### PR TITLE
fix xpu build failure

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118613
* #118528
* #118523
* __->__ #120315

# Motivation
fix build failure introduced by [[DeviceIndex][6/N] Use DeviceIndex in more places](https://github.com/pytorch/pytorch/pull/120133), parameter `total` is undefined in line 100. see https://github.com/pytorch/pytorch/pull/120133/files#diff-00eb8a6f5dfbc341ee9ab9aff0e3dbece8ad73483d4f41a005b1f453cb78221cR91-L102
[PR120133](https://github.com/pytorch/pytorch/pull/120133) forgot to add the label `ciflow/xpu`, so the XPU CI flow was not triggered.

# Solution
refer to [Why is std::cout not printing the correct value for my int8_t number?](https://stackoverflow.com/questions/7587782) , static cast int8_t to int16_t and the condition `device >= 0 && device < total` is enough.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: c10/xpu/XPUFunctions.cpp [hunk 1]
```diff
@@ -90,11 +90,9 @@ void initDeviceProperties(DeviceProp* device_prop, int device) {
 }
 
 inline void check_device(DeviceIndex device) {
+  auto total = static_cast<int16_t>(gDevicePool.devices.size());
```
Comment (guangyey) [inline c10/xpu/XPUFunctions.cpp:93]: refer to [Why is std::cout not printing the correct value for my int8_t number?](https://stackoverflow.com/questions/7587782) , static cast int8_t to int16_t and the condition `device >= 0 && device < total` is enough.

Comment (albanD) [reply c10/xpu/XPUFunctions.cpp:93]: FYI This will conflict with https://github.com/pytorch/pytorch/pull/119639

Comment (guangyey) [reply c10/xpu/XPUFunctions.cpp:93]: Could you help land this PR first? This blocks the XPU build. I will refine this XPU code after #119639 is merged.

Comment (albanD) [reply c10/xpu/XPUFunctions.cpp:93]: fingers crossed it merges cleanly and we can just rebase this PR here.
If it fails for something non trivial, we can merge this one and rebase the other.

Also if you can write this code in a way that it won't logically conflict (works for either type), then we can land them in parallel.

FYI @tringwald

Comment (guangyey) [reply c10/xpu/XPUFunctions.cpp]: Thanks for your information. I change the logic and add TODO comment.

**Review by Skylion007 — APPROVED** [review]

**Review by cyyever — APPROVED** [review]

**Review by malfet — APPROVED** [review]
LGTM, would be good to add some more info why it can happen

**Review by EikanWang — APPROVED** [review]

**Review by gujinghui — APPROVED** [review]

**Comment by guangyey** [comment]
> LGTM, would be good to add some more info why it can happen

Thanks, I enriched the TODO comment.

**Pull request closed without merging**